akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -128,44 +138,182 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def cachedHostConnectionPool[T](host: String, port: Int = 80,
+                                  options: immutable.Traversable[Inet.SocketOption] = Nil,
+                                  settings: Option[HostConnectionPoolSettings] = None,
+                                  defaultHeaders: List[HttpHeader] = Nil,
+                                  log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), HostConnectionPool] = {
+    val effectiveSettings = HostConnectionPoolSettings(settings)
+    val cps = ConnectionPoolSetup(encrypted = false, options, effectiveSettings, defaultHeaders, log)
+    val setup = HostConnectionPoolSetup(host, port, cps)
+    cachedHostConnectionPool(setup)
+  }
+
+  import ConnectionPool.{ HostPool, CowboyEntry }
+
+  // every ActorSystem maintains its own connection pools
+  private[this] val hostPoolCache = new ConcurrentHashMap[HostConnectionPoolSetup, HostPool]
+
+  /**
+   * Returns a [[Flow]] which, when materialized, attaches itself to the per-ActorSystem pool of outgoing
+   * HTTP connections to the given target host endpoint. For every ActorSystem, target host and pool
+   * configuration a separate connection pool is maintained.
+   *
+   * Since the underlying transport usually comprises more than a single connection the produced flow might generate
+   * responses in an order that doesn't directly match the consumed requests.
+   * For example, if two requests A and B enter the flow in that order the response for B might be produced before the
+   * response for A.
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def cachedHostConnectionPool[T](setup: HostConnectionPoolSetup)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), HostConnectionPool] = {
+    @tailrec def hostPool(): HostPool =
+      hostPoolCache.putIfAbsent(setup, CowboyEntry) match {
+        case null ⇒ // we are the first to attempt construction of the HostPool for this setup
+          try {
+            val hostPool = ConnectionPool.hostPool(setup)
+            hostPoolCache.put(setup, hostPool)
+            // we remove the cache entry when the pool has shut down; however, due to the inherent raciness, a client
+            // might still end up receiving a client flow for a pool which is in the process of shutting down or already
+            // dead; in this case the returned ClientFlow will always be completed immediately after materialization
+            hostPool.isShutdown.onComplete(_ ⇒ hostPoolCache.remove(setup))(system.dispatcher)
```

Comment:
the Future should probably be named `whenShutdown`, `isShutdown` sounds a little boolean to me :-)